tmk_core/common/action_util.c

```diff
@@ -364,3 +365,120 @@ __attribute__((weak)) void oneshot_layer_changed_kb(uint8_t layer) { oneshot_lay
  * FIXME: needs doc
  */
 uint8_t has_anymod(void) { return bitpop(real_mods); }
+
+/** \brief custom modifier handling
+ *
+ * This function checks if the given modifier mask is set and the given layer is
+ * the current spec. If not it returns true. If the mask is set and the key
+ * pressed, it deregisters the currently held modifiers, executes the given
+ * function and returns the value returned by the function. If the key is
+ * released, it simply runs the function and returns its value.
+ *
+ * This function is meant to be used in process_record_user to be able to
+ * trigger specific action on key code with modifier, for example when shift-esc
+ * is pressed.
+ *
+ * \param modifier_mask modifier mask to check against current modifiers held.
+ * \param layer layer to check against current layer.
+ * \param keycode the processed keycode, will be forwarded to handler.
+ * \param record the processed record, will be forwarded to handler.
+ * \param data anonymous data to be passed to the handler.
+ * \param handler to be called when modifiers and layer are matched.
+ * \return true if modifiers or layer did not match, return value of handler otherwise.
+ */
+bool run_on_modifiers(uint8_t modifier_mask, int16_t layer, uint16_t keycode,  keyrecord_t *record,
+                      void *data,  bool (*handler)(uint16_t keycode,  keyrecord_t *record, void *data)) {
+  static bool in_alternate_modifier;
+
+  /* nothing to do if we are not in the right layer */
+  if (layer >= 0 && !layer_state_is(layer))
+    return true;
```

Comment:
This kind of layer state testing can result in stuck keys too, because the layer state between the press and release events might change, therefore the custom handler might not be invoked at the release time as it should. Removing this layer testing should fix the problem, because the source layer cache in the QMK core ensures that the release event is always processed using the same layer that was active when the key was pressed.